attendance.php

```diff
@@ -37,6 +37,33 @@
 $cm = get_coursemodule_from_instance('attendance', $attendance->id, 0, false, MUST_EXIST);
 $course = $DB->get_record('course', array('id' => $cm->course), '*', MUST_EXIST);
 
+// If the randomised code is on grab it.
+if ($attforsession->rotateqrcode==1) {
+    $cookiename = 'attendance_'.$attforsession->id;
+    $secrethash = md5($USER->id.$attforsession->rotateqrcodesecret);
+    $url = new moodle_url('/mod/attendance/view.php', array('id' => $cm->id));
+
+    // Check if cookie is set and verify
+    if (isset($_COOKIE[$cookiename])) {
+        // Check the token
+        if ($secrethash !== $_COOKIE[$cookiename]) {
+            // Flag error
+            print_error('qr_cookie_error', 'mod_attendance', $url);
+        }
+    } else {
+        // Check password
+        $qrpassdatabase = $DB->get_record_sql('SELECT * FROM {attendance_rotate_passwords} WHERE attendanceid = ? AND expirytime > ? ORDER BY expirytime ASC LIMIT 1', ['attendanceid'=>$id, time()], $strictness=IGNORE_MISSING);
```

Comment:
I'd like this to also accept the "previous" password if it's expiry was less than 2 seconds ago - hard code it at 2 seconds for now, but we can look at adding a site level setting to control this - it means that if the server is running slow and takes a bit long to receive the request, the code is still accepted - does that make sense?